File: services/apigateway/src/handlers/health.rs
use axum::{extract::State, response::IntoResponse, Json};
use serde_json::json;

use crate::response::ApiResponse;
use crate::AppState;

pub async fn health_check() -> impl IntoResponse {
    Json(ApiResponse::success(json!({
        "status": "healthy",
        "service": "apigateway"
    })))
}

pub async fn detailed_health(State(app_state): State<AppState>) -> impl IntoResponse {
    let mut health_status = json!({
        "status": "healthy",
        "service": "apigateway",
        "timestamp": chrono::Utc::now().to_rfc3339(),
        "dependencies": {}
    });

    // Check Redis connection
    let redis_status = {
        use crate::redis_client::RedisClientExt;
        match app_state.redis_client.ping_api().await {
            Ok(_pong) => json!({
                "status": "healthy",
                "message": "Redis connection successful"
            }),
            Err(e) => json!({
                "status": "unhealthy",
                "message": format!("Redis error: {}", e)
            }),
        }
    };

    health_status["dependencies"]["redis"] = redis_status;

    // Determine overall health
    let is_healthy = health_status["dependencies"]["redis"]["status"] == "healthy";

    if !is_healthy {
        health_status["status"] = json!("degraded");
    }

    Json(ApiResponse::success(health_status))
}

File: services/apigateway-py/app/routers/health.py
"""Health check endpoints"""

from datetime import datetime
from fastapi import APIRouter, Depends
from app.models import ApiResponse, HealthStatus
from app.utils.redis import RedisClient

router = APIRouter(tags=["health"])


async def get_redis_client():
    """Dependency to get Redis client (will be injected from main)"""
    # This will be overridden in main.py
    pass


@router.get("/health", response_model=ApiResponse)
async def health_check():
    """Basic health check endpoint"""
    return ApiResponse(
        success=True, data={"status": "healthy", "service": "apigateway-py"}
    )


@router.get("/health/detailed", response_model=ApiResponse)
async def detailed_health(redis: RedisClient = Depends(get_redis_client)):
    """Detailed health check with dependency status"""
    health = HealthStatus(
        status="healthy",
        service="apigateway-py",
        timestamp=datetime.utcnow(),
        dependencies={},
    )

    # Check Redis connection
    redis_healthy = await redis.ping()
    health.dependencies["redis"] = {
        "status": "healthy" if redis_healthy else "unhealthy",
        "message": "Redis connection successful"
        if redis_healthy
        else "Redis connection failed",
    }

    # Determine overall health
    if not redis_healthy:
        health.status = "degraded"

    return ApiResponse(success=True, data=health.model_dump())
